### DISCUSSION TITLE
Can't add new Unit into UnitDB in Udunits

### DISCUSSION THREAD

**Discussion by rakpawel** [Unidata/netcdf-java#1340] [Questions and Answers]
Hello!

I'm playing with the Udunits library in Java (I know it's available in Udunits 2, but there are no Java implementations).
I'm trying to handle logarithmic units (e.g. dB, dBW, dBm).

What I've done already is:

```
    StandardUnitFormat parser = StandardUnitFormat.instance();
    final Unit W = parser.parse("W");
    final Unit mW = parser.parse("mW");
    final Unit dB = parser.parse("lg(re 1)");
    final Unit BW = parser.parse("lg(re 1 W)");
    final Unit dBW = parser.parse("0.1 lg(re 1 W)");
    final Unit dBm = parser.parse("0.1 lg(re 1 mW)");
    

    float convertedBWtoW = BW.convertTo(10, W);
    float converted10dBWtoW = dBW.convertTo(10, W);
    float converted15dBWtoW = dBW.convertTo(15, W);
 ```

and it all works nicely. 

What I'm trying to achieve is to create dB prefix and add it to `UnitDB` to be able to parse strings like "10 dBm" later on in my application. 
Perfect scenario would be to have also `dB` added to UnitDB and be able to combine it with all different units.

I can  add `UnitDBManager.instance().addUnit(parser.parse("lg(re 1)"))` - but then no `UnitName` is specified. 

It looks like I'm missing something really obvious, could you point me into the right direction?

**Comment by tdrwenski** [comment]
If I understand correctly, your problem is that the `parser.parse` is returning a nameless unit which can't be added to your database. I am not very familiar with Udunits so there may be a better way, but I believe one solution is to clone the unit and add a name before adding it to your database, like:
```
    Unit namelessUnit = parser.parse("lg(re 1)");
    Unit namedUnit = namelessUnit.clone(UnitName.newUnitName("dB"));
    UnitDBManager.instance().addUnit(namedUnit);
```

**Reply by rakpawel** [reply]
Thank you! I missed that.

What I'm looking for is possibility of defining just "B" or "dB" and having it work out of the box with decibel-watts (dBW) or decibel-milliwatts.

If I do that:
```
Unit BW = parser.parse("lg(re 1 W)").clone(UnitName.newUnitName("bel-watt", "bel-watts", "BW"));
UnitDBManager.instance().addUnit(BW);
```
or that:
```
LogarithmicUnit BW = new LogarithmicUnit(SI.Watt, 10, UnitName.newUnitName("bel-watt", "bel-watts", "BW"));
UnitDBManager.instance().addUnit(BW);
```

Then it works for both watts and mili-watts, but I was wondering If there is a way to define bel/decibel as some kind of logarithmic prefix that will work with watts, volts etc.

I can see that it's defined in UDUNITS-2.
```
<unit>
    <def>lg(re 1)</def>
    <aliases> <symbol>B</symbol> </aliases>
    <definition>logarithmic unit used to express dimensionless ratios; a value of 0 B (0 dB) corresponds to a ratio of 1</definition>
</unit>
 ```

Just found out that UDUNITS-2 defines all my cases separately, so that's probably the way to go :)

```
<unit>
    <def>lg(re 1 W)</def>
    <aliases> <symbol>BW</symbol> </aliases>
    <definition>unit of power in decibel scale, referenced to 1 watt; the power in bel-watts is equal to 100 times the base 10 logarithm of the power in watts (making decibel-watt, the more common term, 10 times the base 10 log of the power in watts)</definition>
</unit>
<unit>
    <def>lg(re 1 mW)</def>
    <aliases> <symbol>Bm</symbol> </aliases>
    <definition>unit of power in decibel scale, referenced to 1 milliwatt; the power in bel-milliwatts is equal to 100 times the base 10 logarithm of the power in milliwatts (making decibel-milliwatt, the more common term, 10 times the base 10 log of the power in milliwatts)</definition>
</unit>
```
